lib/src/source/hosted.dart

```diff
@@ -220,7 +221,11 @@ class BoundHostedSource extends CachedSource {
     try {
       // TODO(sigurdm): Implement cancellation of requests. This probably
       // requires resolution of: https://github.com/dart-lang/sdk/issues/22265.
-      bodyText = await httpClient.read(url, headers: pubApiHeaders);
+      bodyText = await withAuthenticatedClient(
+        systemCache,
+        url.toString(),
+        (client) => client.read(url, headers: pubApiHeaders),
+      );
```

Comment:
So if you use token-authentication for `pub.dev` we'll be adding a user identifiable authentication token to all requests..

Authenticating requests that don't need it could be undesirable.. for multiple reasons:
 * Respecting users privacy by not sending unnecessary user identifiable tokens,
 * Attaching secret tokens to requests that don't need it can only increase the risk we leak them accidentally,
 * We could imagine that some tokens like the oauth stuff used in pub.dev needs to be refreshed, using tokens when not needed could reduce performance,
 * Attaching an unnecessary `Authorization` header is often an effective way to blow through CDN or other web cache... reducing performance further.

I realize that in it's current form `withAuthenticatedClient` isn't intended for use when authenticating to `pub.dartlang.org`, but I imagine that we'll want to support authenticating with tokens here too.

-------------------------------

Maybe, it's acceptable for 3rd party pub servers, since most of them will likely require a token for fetching packages anyways.

In that case we can just specialize `pub.dartlang.org` if/when we add support for `pub.dartlang.org`.

Reply:
We can add scopes to authentication schemes which might be used to determine which request should be / could be authenticated using given scheme. It could be added in future without making breaking change.